src/Symfony/Bundle/FrameworkBundle/CacheWarmer/ControllerMetadataCacheWarmer.php

```diff
@@ -0,0 +1,135 @@
+
+use Symfony\Component\Config\ConfigCacheFactoryInterface;
+use Symfony\Component\Config\ConfigCacheInterface;
+use Symfony\Component\Config\Resource\FileResource;
+use Symfony\Component\Filesystem\Filesystem;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpKernel\CacheWarmer\CacheWarmerInterface;
+use Symfony\Component\HttpKernel\Controller\ControllerResolverInterface;
+use Symfony\Component\HttpKernel\ControllerMetadata\ControllerMetadataFactoryInterface;
+use Symfony\Component\HttpKernel\ControllerMetadata\ControllerMetadataUtil;
+use Symfony\Component\Routing\Route;
+use Symfony\Component\Routing\RouterInterface;
+
+/**
+ * Generates the cache for controller metadata.
+ *
+ * @author Iltar van der Berg <kjarli@gmail.com>
+ */
+final class ControllerMetadataCacheWarmer implements CacheWarmerInterface
+{
+    private $router;
+    private $controllerResolver;
+    private $controllerMetadataFactory;
+    private $configCacheFactory;
+    private $cacheFile;
+
+    public function __construct(RouterInterface $router, ControllerResolverInterface $controllerResolver, ConfigCacheFactoryInterface $configCacheFactory, ControllerMetadataFactoryInterface $controllerMetadataFactory, $cacheFile)
+    {
+        $this->router = $router;
+        $this->controllerResolver = $controllerResolver;
+        $this->configCacheFactory = $configCacheFactory;
+        $this->controllerMetadataFactory = $controllerMetadataFactory;
+        $this->cacheFile = $cacheFile;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function warmUp($cacheDir)
+    {
+        $this->configCacheFactory->cache($this->cacheFile, function (ConfigCacheInterface $cache) {
+            $this->refreshCache($cache);
+        });
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function isOptional()
+    {
+        return true;
+    }
+
+    private function refreshCache(ConfigCacheInterface $cache)
+    {
+        // make sure the cache is not used when executing createControllerMetadata
+        (new Filesystem())->remove($cache->getPath());
+
+        $controllers = array();
+
+        foreach ($this->router->getRouteCollection()->all() as $route) {
+            /* @var $route Route */
+            if (!$route->hasDefault('_controller')) {
+                continue;
+            }
+
+            $controller = $this->controllerResolver->getController(new Request(array(), array(), array('_controller' => $route->getDefault('_controller'))));
+
+            if (null === ($logicalName = ControllerMetadataUtil::getControllerLogicalName($controller))) {
+                continue;
+            }
+
+            list($className, $methodName) = $logicalName;
+
+            if (null === $className || isset($controllers[$className])) {
+                // either a closure or controller is already processed
+                continue;
+            }
+
+            $controllers[$className][$methodName] = true;
```

Comment:
I wonder, if we should care about such applications?
I mean, if someone use weird methods or magic, the cache will not be warmed and routes will be loaded with the standard way (right?). So should we add complexity to handle those cases?
If no, we may remove the reflexion and just use the method defined in the route